i6_native_ops/fbw/__init__.py

```diff
@@ -1,2 +1,68 @@
+import os
 import torch  # needed to find pytorch specific libs
-from .fbw_core import fbw
+from pkg_resources import get_distribution
+from typing import Tuple
+
+try:
+    # Package is installed, so ops are already compiled
+    __version__ = get_distribution('i6_native_ops').version
+    from . import fbw_core as core
+except Exception as e:
+    # otherwise try to build locally
+    from torch.utils.cpp_extension import load
+    base_path = os.path.dirname(__file__)
+    core = load(
+        name="fbw_core",
+        sources=[
+            os.path.join(base_path, "fbw_torch.cpp"),
+            os.path.join(base_path, "fbw_op.cu"),
+        ],
+        extra_include_paths=[
+            base_path,
+            os.path.join(base_path, "..", "common")
+        ],
+    )
+
+
+class FastBaumWelchLoss(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, am_scores, fsa, seq_lens):
+        num_states, edge_tensor, weight_tensor, start_end_states = fsa
+
+        grad, loss = core.fbw(
+            am_scores, edge_tensor, weight_tensor,
+            start_end_states, seq_lens, int(num_states),
+            core.DebugOptions()
+        )
+        ctx.save_for_backward(grad)
+        return loss
+    
+    @staticmethod
+    def backward(ctx, grad_loss):
+        # negative log prob -> prob
+        grad = ctx.saved_tensors[0].neg().exp()
+        return grad, None, None
+
+
+def fbw_loss(
+    log_probs: torch.FloatTensor,
+    fsa: Tuple[int, torch.IntTensor, torch.FloatTensor, torch.IntTensor],
+    seq_lens: torch.IntTensor
+) -> Tuple[torch.IntTensor, torch.FloatTensor]:
```

Comment:
The return type seems wrong to me, is this not only one element?

Reply:
Yes you're right